src/Umbraco.Core/Persistence/Repositories/VersionableRepositoryBase.cs

```diff
@@ -454,7 +454,6 @@ INNER JOIN
 
                             //this property has tags, so we need to extract them and for that we need the prevals which we've already looked up
                             var preValData = allPreValues.Value.Where(x => x.DataTypeNodeId == property.PropertyType.DataTypeDefinitionId)
-                                .Distinct()

```

Comment:
Not sure if this is the correct fix, if Distinct() isn't actually returning distinct values correctly, the the underyling issue is probably that DataTypePreValueDto hasn't implemented Equals or GetHashCode correctly.
Currently I can see that it is only using 'Id' to do this equals/hash check which might be incorrect.... since as you say it is actually filtering all of the prevalues for this data type.